src/blocks/battery/apc_ups.rs

```diff
@@ -0,0 +1,196 @@
+        } else {
+            return Err(Error::new(format!(
+                "Expected unit for {} are {}, but got {}",
+                property_name, required_unit, unit
+            )));
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ApcConnection(TcpStream);
+
+impl ApcConnection {
+    async fn connect(addr: String) -> Result<Self> {
+        Ok(Self(
+            TcpStream::connect(addr)
+                .await
+                .error("Failed to connect to socket")?,
+        ))
+    }
+
+    async fn write(&mut self, msg: &[u8]) -> Result<()> {
+        let msg_len = u16::try_from(msg.len())
+            .error("msg is too long, it must be less than 2^16 characters long")?;
+
+        self.0
+            .write_u16(msg_len)
+            .await
+            .error("Could not write message length to socket")?;
+        self.0
+            .write_all(msg)
+            .await
+            .error("Could not write message to socket")?;
+        Ok(())
+    }
+
+    async fn read_response(&mut self) -> Result<String> {
+        let mut buf = String::new();
+        loop {
+            let read_size = self
+                .0
+                .read_u16()
+                .await
+                .error("Could not read response length from socket")?
+                .into();
+            if read_size == 0 {
+                break;
+            }
+            let mut read_buf = vec![0_u8; read_size];
+            self.0
+                .read_exact(&mut read_buf)
+                .await
+                .error("Could not read from socket")?;
+            buf.extend(String::from_utf8(read_buf));
+        }
+        Ok(buf)
+    }
+}
+
+pub(super) struct Device {
+    addr: String,
+    interval: Interval,
+}
+
+impl Device {
+    pub(super) async fn new(dev_name: DeviceName, interval: Seconds) -> Result<Self> {
+        let addr = dev_name.exact().unwrap_or("localhost:3551");
+        Ok(Self {
+            addr: addr.to_string(),
+            interval: interval.timer(),
+        })
+    }
+
+    async fn get_status(&mut self) -> Result<PropertyMap> {
+        let mut conn = ApcConnection::connect(self.addr.clone()).await?;
+
+        conn.write(b"status").await?;
+        let response = conn.read_response().await?;
+
+        let mut property_map = PropertyMap::default();
+
+        for line in response.lines() {
+            if let Some((key, value)) = line.split_once(':') {
+                property_map.insert(key.trim().to_string(), value.trim().to_string());
+            }
+        }
+
+        Ok(property_map)
+    }
+}
+
+#[async_trait]
+impl BatteryDevice for Device {
+    async fn get_info(&mut self) -> Result<Option<BatteryInfo>> {
+        let status_data = self
+            .get_status()
+            .await
+            .map_err(|e| {
+                debug!("{e}");
+                e
+            })
+            .unwrap_or_default();
+
+        let status_str = status_data.get("STATUS").map_or("COMMLOST", |x| x);
```

Comment:
@MaxVerevkin , any tips on writing this or the previous let more cleanly?

Reply:
I think it's good :)

Reply:
Ok the `|x| x` felt kinda hacky to get from the `&String` to a `&str`, but if you think it's kosher, I'm happy!

Reply:
I'm happy either way :)

Reply:
changed in my commit